File: src/api/aria2.ts
/** @fileoverview Aria2 JSON-RPC client wrapper providing typed API for task management. */
import { Aria2 } from '@shared/aria2'
import { TASK_STATUS } from '@shared/constants'
import {
    changeKeysToCamelCase,
    formatOptionsForEngine,
} from '@shared/utils'

let client: Aria2 | null = null
let engineReady = false

/** Returns true when the aria2 RPC connection has been established. */
export function isEngineReady(): boolean {
    return engineReady
}

/** Returns the initialized Aria2 client instance or throws if not yet initialized. */
export function getClient(): Aria2 {
    if (!client) throw new Error('Aria2 client not initialized')
    return client
}

/** Creates and opens a new Aria2 RPC client connection. */
export async function initClient(options: { port: number; secret: string }) {
    client = new Aria2({
        host: '127.0.0.1',
        port: options.port,
        secret: options.secret,
    })
    await client.open()
    engineReady = true
    return client
}

/** Closes the active Aria2 RPC connection and resets the client. */
export async function closeClient() {
    if (client) {
        await client.close()
        client = null
    }
}

/** Retrieves aria2 engine version and list of enabled features. */
export async function getVersion() {
    return getClient().call('getVersion') as Promise<{ version: string; enabledFeatures: string[] }>
}

export async function getGlobalOption() {
    const data = await getClient().call('getGlobalOption')
    return changeKeysToCamelCase(data as Record<string, unknown>)
}

export async function getGlobalStat() {
    return getClient().call('getGlobalStat') as Promise<Record<string, string>>
}

export async function changeGlobalOption(options: Record<string, unknown>) {
    const engineOptions = formatOptionsForEngine(options)
    return getClient().call('changeGlobalOption', engineOptions)
}

export async function getOption(params: { gid: string }) {
    const data = await getClient().call('tellStatus', params.gid)
    return changeKeysToCamelCase(data as Record<string, unknown>)
}

export async function changeOption(params: { gid: string; options: Record<string, unknown> }) {
    const engineOptions = formatOptionsForEngine(params.options)
    return getClient().call('changeOption', params.gid, engineOptions)
}

async function tellStatus(gid: string) {
    return getClient().call('tellStatus', gid) as Promise<Record<string, unknown>>
}

async function tellActive() {
    return getClient().call('tellActive') as Promise<Record<string, unknown>[]>
}

async function tellWaiting(offset: number, num: number) {
    return getClient().call('tellWaiting', offset, num) as Promise<Record<string, unknown>[]>
}

async function tellStopped(offset: number, num: number) {
    return getClient().call('tellStopped', offset, num) as Promise<Record<string, unknown>[]>
}

export async function fetchActiveTaskList() {
    return tellActive()
}

export async function fetchTaskList(params: { type: string }) {
    const { type } = params
    switch (type) {
        case TASK_STATUS.ACTIVE: {
            const [active, waiting] = await Promise.all([
                tellActive(),
                tellWaiting(0, 1000),
            ])
            return [...active, ...waiting] as Record<string, unknown>[]
        }
        default:
            return tellStopped(0, 1000)
    }
}

export async function fetchTaskItem(params: { gid: string }) {
    return tellStatus(params.gid)
}

export async function fetchTaskItemWithPeers(params: { gid: string }) {
    const [task, peers] = await Promise.all([
        tellStatus(params.gid),
        getClient().call('getPeers', params.gid) as Promise<unknown[]>,
    ])
    return { ...task, peers }
}

/** Adds one or more URI downloads with per-URI output filename overrides. */
export async function addUri(params: { uris: string[]; outs: string[]; options: Record<string, unknown> }) {
    const { uris, outs, options } = params
    const engineOptions = formatOptionsForEngine(options)
    const tasks = uris.map((uri, index) => {
        const opts = { ...engineOptions }
        if (outs[index]) opts.out = outs[index]
        return getClient().call('addUri', [uri], opts)
    })
    return Promise.all(tasks)
}

/** Adds a torrent download from a base64-encoded .torrent file. */
export async function addTorrent(params: { torrent: string; options: Record<string, unknown> }) {
    const engineOptions = formatOptionsForEngine(params.options)
    return getClient().call('addTorrent', params.torrent, [], engineOptions)
}

export async function addMetalink(params: { metalink: string; options: Record<string, unknown> }) {
    const engineOptions = formatOptionsForEngine(params.options)
    return getClient().call('addMetalink', params.metalink, engineOptions)
}

export async function removeTask(params: { gid: string }) {
    return getClient().call('forceRemove', params.gid)
}

export async function forcePauseTask(params: { gid: string }) {
    return getClient().call('forcePause', params.gid)
}

export async function pauseTask(params: { gid: string }) {
    return getClient().call('pause', params.gid)
}

export async function resumeTask(params: { gid: string }) {
    return getClient().call('unpause', params.gid)
}

export async function pauseAllTask() {
    return getClient().call('pauseAll')
}

export async function forcePauseAllTask() {
    return getClient().call('forcePauseAll')
}

export async function resumeAllTask() {
    return getClient().call('unpauseAll')
}

export async function saveSession() {
    return getClient().call('saveSession')
}

export async function removeTaskRecord(params: { gid: string }) {
    return getClient().call('removeDownloadResult', params.gid)
}

export async function purgeTaskRecord() {
    return getClient().call('purgeDownloadResult')
}

export async function batchResumeTask(params: { gids: string[] }) {
    const calls = params.gids.map((gid) => ['unpause', gid] as [string, ...unknown[]])
    return getClient().multicall(calls)
}

export async function batchPauseTask(params: { gids: string[] }) {
    const calls = params.gids.map((gid) => ['forcePause', gid] as [string, ...unknown[]])
    return getClient().multicall(calls)
}

export async function batchForcePauseTask(params: { gids: string[] }) {
    return batchPauseTask(params)
}

export async function batchRemoveTask(params: { gids: string[] }) {
    const calls = params.gids.map((gid) => ['forceRemove', gid] as [string, ...unknown[]])
    return getClient().multicall(calls)
}

/** Fetches user preferences from the Tauri persistent store. */
export async function fetchPreference() {
    const { invoke } = await import('@tauri-apps/api/core')
    return invoke('get_app_config') as Promise<Record<string, unknown>>
}

/** Saves updated user preferences to the Tauri persistent store. */
export async function savePreference(config: Record<string, unknown>) {
    const { invoke } = await import('@tauri-apps/api/core')
    return invoke('save_preference', { config })
}

const api = {
    initClient,
    closeClient,
    getVersion,
    getGlobalOption,
    getGlobalStat,
    changeGlobalOption,
    getOption,
    changeOption,
    fetchActiveTaskList,
    fetchTaskList,
    fetchTaskItem,
    fetchTaskItemWithPeers,
    addUri,
    addTorrent,
    addMetalink,
    removeTask,
    forcePauseTask,
    pauseTask,
    resumeTask,
    pauseAllTask,
    forcePauseAllTask,
    resumeAllTask,
    saveSession,
    removeTaskRecord,
    purgeTaskRecord,
    batchResumeTask,
    batchPauseTask,
    batchForcePauseTask,
    batchRemoveTask,
    fetchPreference,
    savePreference,
}

export default api

File: src-tauri/src/commands/aria2.rs
//! Tauri commands exposing aria2 RPC operations to the frontend.
//!
//! These commands serve as the invoke() transport layer. Each command maps
//! to one or more aria2 RPC methods.

use crate::aria2::client::Aria2State;
use crate::aria2::types::{Aria2File, Aria2Task};
use crate::error::AppError;
use tauri::State;

/// Fetch task list by type: "active" returns active+waiting, otherwise stopped.
#[tauri::command]
pub async fn aria2_fetch_task_list(
    state: State<'_, Aria2State>,
    r#type: String,
    limit: Option<i64>,
) -> Result<Vec<Aria2Task>, AppError> {
    if r#type == "active" {
        let (active, waiting) =
            tokio::try_join!(state.0.tell_active(), state.0.tell_waiting(0, 1000),)?;
        let mut result = active;
        result.extend(waiting);
        Ok(result)
    } else {
        state.0.tell_stopped(0, limit.unwrap_or(1000)).await
    }
}

/// Fetch only active tasks (no waiting).
#[tauri::command]
pub async fn aria2_fetch_active_task_list(
    state: State<'_, Aria2State>,
) -> Result<Vec<Aria2Task>, AppError> {
    state.0.tell_active().await
}

/// Fetch a single task's full status by GID.
#[tauri::command]
pub async fn aria2_fetch_task_item(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<Aria2Task, AppError> {
    state.0.tell_status(&gid).await
}

/// Fetch task status with peer list (for BT tasks).
#[tauri::command]
pub async fn aria2_fetch_task_item_with_peers(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<serde_json::Value, AppError> {
    let (task, peers) = tokio::try_join!(state.0.tell_status(&gid), state.0.get_peers(&gid),)?;
    let mut result =
        serde_json::to_value(&task).map_err(|e| AppError::Aria2(format!("serialize task: {e}")))?;
    result["peers"] = peers;
    Ok(result)
}

/// Get aria2 engine version and enabled features.
#[tauri::command]
pub async fn aria2_get_version(
    state: State<'_, Aria2State>,
) -> Result<serde_json::Value, AppError> {
    state.0.get_version().await
}

/// Get global aria2 options.
#[tauri::command]
pub async fn aria2_get_global_option(
    state: State<'_, Aria2State>,
) -> Result<serde_json::Value, AppError> {
    state.0.get_global_option().await
}

/// Get global download/upload statistics.
#[tauri::command]
pub async fn aria2_get_global_stat(
    state: State<'_, Aria2State>,
) -> Result<serde_json::Value, AppError> {
    let stat = state.0.get_global_stat().await?;
    serde_json::to_value(&stat).map_err(|e| AppError::Aria2(format!("serialize stat: {e}")))
}

/// Change global aria2 options at runtime.
#[tauri::command]
pub async fn aria2_change_global_option(
    state: State<'_, Aria2State>,
    options: serde_json::Map<String, serde_json::Value>,
) -> Result<String, AppError> {
    state.0.change_global_option(options).await
}

/// Get per-task options.
#[tauri::command]
pub async fn aria2_get_option(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<serde_json::Value, AppError> {
    state.0.get_option(&gid).await
}

/// Change per-task options.
#[tauri::command]
pub async fn aria2_change_option(
    state: State<'_, Aria2State>,
    gid: String,
    options: serde_json::Value,
) -> Result<String, AppError> {
    state.0.change_option(&gid, options).await
}

/// Get file list for a task.
#[tauri::command]
pub async fn aria2_get_files(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<Vec<Aria2File>, AppError> {
    state.0.get_files(&gid).await
}

// ── `out` option sanitization ────────────────────────────────────────

/// Extracts the bare filename from an `out` option value.
///
/// aria2's `out` option must be a plain filename relative to `dir`.
/// This is a **basename extraction** helper — it strips path components
/// (including Windows drive letters, UNC prefixes, and Unix absolute
/// paths) but does NOT perform full Windows filename sanitization
/// (reserved names, illegal characters).  Those are left to aria2's
/// own validation.
///
/// Returns `None` for values that reduce to empty, `.`, `..`, or
/// contain NUL bytes (which would truncate C strings inside aria2).
fn sanitize_out_option(raw: &str) -> Option<&str> {
    if raw.is_empty() {
        return None;
    }
    // Split on both separators to handle cross-platform paths.
    // rsplit + next always returns Some for non-empty input.
    let basename = raw.rsplit(|c| c == '/' || c == '\\').next().unwrap_or(raw);
    if basename.is_empty() || basename == "." || basename == ".." {
        return None;
    }
    if basename.contains('\0') {
        return None;
    }
    Some(basename)
}

/// Add URI download(s). Each URI gets its own aria2 task with optional
/// per-URI `out` filename override and file-category directory resolution.
#[tauri::command]
pub async fn aria2_add_uri(
    state: State<'_, Aria2State>,
    uris: Vec<String>,
    mut options: serde_json::Value,
) -> Result<String, AppError> {
    // Enforce out = filename-only invariant before forwarding to aria2.
    // Prevents doubled paths when `out` accidentally contains an absolute
    // path (e.g., from session restore or unvalidated external input). (#261)
    if let Some(opts) = options.as_object_mut() {
        if let Some(out_val) = opts.get("out").and_then(|v| v.as_str()).map(String::from) {
            match sanitize_out_option(&out_val) {
                Some(clean) if clean != out_val => {
                    log::warn!(
                        "aria2:add-uri sanitized out: original_had_path=true clean={clean:?}"
                    );
                    opts.insert(
                        "out".to_string(),
                        serde_json::Value::String(clean.to_string()),
                    );
                }
                None => {
                    log::warn!("aria2:add-uri removed invalid out option");
                    opts.remove("out");
                }
                _ => {} // already a clean filename — no action needed
            }
        }
    }
    log::info!("aria2:add-uri count={}", uris.len());
    state.0.add_uri(uris, options).await
}

/// Add a torrent download from base64-encoded content.
#[tauri::command]
pub async fn aria2_add_torrent(
    state: State<'_, Aria2State>,
    torrent: String,
    options: serde_json::Value,
) -> Result<String, AppError> {
    log::info!("aria2:add-torrent");
    state.0.add_torrent(&torrent, options).await
}

/// Add a metalink download from base64-encoded content.
#[tauri::command]
pub async fn aria2_add_metalink(
    state: State<'_, Aria2State>,
    metalink: String,
    options: serde_json::Value,
) -> Result<Vec<String>, AppError> {
    log::info!("aria2:add-metalink");
    state.0.add_metalink(&metalink, options).await
}

/// Forcefully remove a task by GID.
#[tauri::command]
pub async fn aria2_force_remove(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<String, AppError> {
    log::info!("aria2:remove gid={gid}");
    state.0.force_remove(&gid).await
}

/// Forcefully pause a task by GID.
#[tauri::command]
pub async fn aria2_force_pause(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<String, AppError> {
    log::debug!("aria2:force-pause gid={gid}");
    state.0.force_pause(&gid).await
}

/// Gracefully pause a task.
#[tauri::command]
pub async fn aria2_pause(state: State<'_, Aria2State>, gid: String) -> Result<String, AppError> {
    log::debug!("aria2:pause gid={gid}");
    state.0.pause(&gid).await
}

/// Resume a paused task.
#[tauri::command]
pub async fn aria2_unpause(state: State<'_, Aria2State>, gid: String) -> Result<String, AppError> {
    log::debug!("aria2:resume gid={gid}");
    state.0.unpause(&gid).await
}

/// Pause all active downloads (graceful).
#[tauri::command]
pub async fn aria2_pause_all(state: State<'_, Aria2State>) -> Result<String, AppError> {
    state.0.pause_all().await
}

/// Forcefully pause all active downloads.
#[tauri::command]
pub async fn aria2_force_pause_all(state: State<'_, Aria2State>) -> Result<String, AppError> {
    state.0.force_pause_all().await
}

/// Resume all paused downloads.
#[tauri::command]
pub async fn aria2_unpause_all(state: State<'_, Aria2State>) -> Result<String, AppError> {
    state.0.unpause_all().await
}

/// Save the current aria2 session to disk.
#[tauri::command]
pub async fn aria2_save_session(state: State<'_, Aria2State>) -> Result<String, AppError> {
    state.0.save_session().await
}

/// Remove a completed/errored task record from aria2's download list.
#[tauri::command]
pub async fn aria2_remove_download_result(
    state: State<'_, Aria2State>,
    gid: String,
) -> Result<String, AppError> {
    state.0.remove_download_result(&gid).await
}

/// Purge all completed/errored download results.
#[tauri::command]
pub async fn aria2_purge_download_result(state: State<'_, Aria2State>) -> Result<String, AppError> {
    log::info!("aria2:purge-results");
    state.0.purge_download_result().await
}

/// Batch resume multiple tasks via multicall.
#[tauri::command]
pub async fn aria2_batch_unpause(
    state: State<'_, Aria2State>,
    gids: Vec<String>,
) -> Result<Vec<serde_json::Value>, AppError> {
    log::info!("aria2:batch-resume count={}", gids.len());
    let calls = gids
        .into_iter()
        .map(|gid| ("unpause".to_string(), vec![serde_json::Value::String(gid)]))
        .collect();
    state.0.multicall(calls).await
}

/// Batch force-pause multiple tasks via multicall.
#[tauri::command]
pub async fn aria2_batch_force_pause(
    state: State<'_, Aria2State>,
    gids: Vec<String>,
) -> Result<Vec<serde_json::Value>, AppError> {
    log::info!("aria2:batch-pause count={}", gids.len());
    let calls = gids
        .into_iter()
        .map(|gid| {
            (
                "forcePause".to_string(),
                vec![serde_json::Value::String(gid)],
            )
        })
        .collect();
    state.0.multicall(calls).await
}

/// Batch force-remove multiple tasks via multicall.
#[tauri::command]
pub async fn aria2_batch_force_remove(
    state: State<'_, Aria2State>,
    gids: Vec<String>,
) -> Result<Vec<serde_json::Value>, AppError> {
    log::info!("aria2:batch-remove count={}", gids.len());
    let calls = gids
        .into_iter()
        .map(|gid| {
            (
                "forceRemove".to_string(),
                vec![serde_json::Value::String(gid)],
            )
        })
        .collect();
    state.0.multicall(calls).await
}

#[cfg(test)]
mod tests {
    use super::sanitize_out_option;

    #[test]
    fn bare_filename_passes_through() {
        assert_eq!(sanitize_out_option("file.zip"), Some("file.zip"));
    }

    #[test]
    fn windows_backslash_absolute_extracts_basename() {
        assert_eq!(
            sanitize_out_option("C:\\Users\\u\\Downloads\\file.zip"),
            Some("file.zip")
        );
    }

    #[test]
    fn forward_slash_absolute_extracts_basename() {
        assert_eq!(
            sanitize_out_option("C:/Users/u/Downloads/file.zip"),
            Some("file.zip")
        );
    }

    #[test]
    fn unc_path_extracts_basename() {
        assert_eq!(
            sanitize_out_option("\\\\server\\share\\file.zip"),
            Some("file.zip")
        );
    }

    // basename extraction: "../evil.exe" → "evil.exe".
    // This strips the traversal prefix; it is NOT a security path-join
    // guard — aria2 resolves the final path via applyDir(dir, out).
    #[test]
    fn parent_traversal_extracts_basename() {
        assert_eq!(sanitize_out_option("../evil.exe"), Some("evil.exe"));
    }

    #[test]
    fn dotdot_only_rejected() {
        assert_eq!(sanitize_out_option(".."), None);
    }

    #[test]
    fn dot_only_rejected() {
        assert_eq!(sanitize_out_option("."), None);
    }

    #[test]
    fn empty_rejected() {
        assert_eq!(sanitize_out_option(""), None);
    }

    #[test]
    fn nul_byte_rejected() {
        assert_eq!(sanitize_out_option("file\0.zip"), None);
    }

    #[test]
    fn cjk_filename_preserved() {
        assert_eq!(sanitize_out_option("C:/下载/文件.zip"), Some("文件.zip"));
    }

    #[test]
    fn trailing_separator_rejected() {
        assert_eq!(sanitize_out_option("path/to/"), None);
    }

    // Exact regression test for issue #261: category dir = Downloads/Programs,
    // out contained full base dir path → applyDir doubled it.
    #[test]
    fn issue_261_regression() {
        assert_eq!(
            sanitize_out_option("C:/Users/37472/Downloads/sysdiag-all-x64.exe"),
            Some("sysdiag-all-x64.exe")
        );
    }
}
